crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/ManifestPuller.java

```diff
@@ -103,4 +94,38 @@ public String getActionDescription() {
         + ":"
         + imageTag;
   }
+
+  /**
+   * Instantiates a {@link ManifestTemplate} from a JSON string. This checks the {@code
+   * schemaVersion} field of the JSON to determine which manifest version to use.
+   */
+  private T getManifestTemplateFromJson(String jsonString)
+      throws IOException, UnknownManifestFormatException {
+    ObjectNode node = new ObjectMapper().readValue(jsonString, ObjectNode.class);
+    if (!node.has("schemaVersion")) {
+      throw new UnknownManifestFormatException("Cannot find field 'schemaVersion' in manifest");
+    }
+
+    if (!manifestTemplateClass.equals(ManifestTemplate.class)) {
+      return JsonTemplateMapper.readJson(jsonString, manifestTemplateClass);
+    }
+
+    int schemaVersion = node.get("schemaVersion").asInt(-1);
+    switch (schemaVersion) {
+      case 1:
+        return manifestTemplateClass.cast(
+            JsonTemplateMapper.readJson(jsonString, V21ManifestTemplate.class));
+
+      case 2:
+        return manifestTemplateClass.cast(
+            JsonTemplateMapper.readJson(jsonString, V22ManifestTemplate.class));
+
+      case -1:
```

Comment:
if else would likely be more obvious; especially since your cases are out of order which interferes with the compiler optimizations for switch